linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -0,0 +1,3241 @@
+          case _ =>
+            fb += wa.Call(genFunctionID.unbox(targetTpe.primRef))
+        }
+    }
+  }
+
+  private def genGetClass(tree: GetClass): Type = {
+    /* Unlike in `genApply` or `genStringConcat`, here we make no effort to
+     * optimize known-primitive receivers. In practice, such cases would be
+     * useless.
+     */
+
+    val needHijackedClassDispatch = tree.expr.tpe match {
+      case ClassType(className) =>
+        ctx.getClassInfo(className).isAncestorOfHijackedClass
+      case ArrayType(_) | NothingType | NullType =>
+        false
+      case _ =>
+        true
+    }
+
+    if (!needHijackedClassDispatch) {
+      val typeDataType = watpe.RefType(genTypeID.typeData)
+      val objectTypeIdx = genTypeID.forClass(ObjectClass)
+
+      val typeDataLocal = addSyntheticLocal(typeDataType)
+
+      genTreeAuto(tree.expr)
+      markPosition(tree)
+      fb += wa.StructGet(objectTypeIdx, genFieldID.objStruct.vtable) // implicit trap on null
+      fb += wa.LocalSet(typeDataLocal)
+      genClassOfFromTypeData(wa.LocalGet(typeDataLocal))
+    } else {
+      genTree(tree.expr, AnyType)
+      markPosition(tree)
+      fb += wa.RefAsNonNull
+      fb += wa.Call(genFunctionID.anyGetClass)
+    }
+
+    tree.tpe
+  }
+
+  private def genReadStorage(storage: VarStorage): Unit = {
+    storage match {
+      case VarStorage.Local(localID) =>
+        fb += wa.LocalGet(localID)
+      case VarStorage.StructField(structLocal, structTypeID, fieldID) =>
+        fb += wa.LocalGet(structLocal)
+        fb += wa.StructGet(structTypeID, fieldID)
+    }
+  }
+
+  private def genVarRef(r: VarRef): Type = {
+    markPosition(r)
+    if (r.tpe == NothingType)
+      fb += wa.Unreachable
+    else
+      genReadStorage(lookupLocal(r.ident.name))
+    r.tpe
+  }
+
+  private def genThis(t: This): Type = {
+    markPosition(t)
+    genReadStorage(receiverStorage)
+    t.tpe
+  }
+
+  private def genVarDef(r: VarDef): Type = {
+    /* This is an isolated VarDef that is not in a Block.
+     * Its scope is empty by construction, and therefore it need not be stored.
+     */
+    genTree(r.rhs, NoType)
+    NoType
+  }
+
+  private def genIf(t: If, expectedType: Type): Type = {
+    val ty = transformResultType(expectedType)
+    genTree(t.cond, BooleanType)
+
+    markPosition(t)
+
+    t.elsep match {
+      case Skip() =>
+        assert(expectedType == NoType)
+        fb.ifThen() {
+          genTree(t.thenp, expectedType)
+        }
+      case _ =>
+        fb.ifThenElse(ty) {
+          genTree(t.thenp, expectedType)
+        } {
+          genTree(t.elsep, expectedType)
+        }
+    }
+
+    if (expectedType == NothingType)
+      fb += wa.Unreachable
+
+    expectedType
+  }
+
+  private def genWhile(t: While): Type = {
+    t.cond match {
+      case BooleanLiteral(true) =>
+        // infinite loop that must be typed as `nothing`, i.e., unreachable
+        markPosition(t)
```

Comment:
I don't think this helps in this case, because of the additional `markPosition` in the middle, which cannot be factored out. Moving this one up makes it less obvious that everything is well-positioned, IMO.